File: intrinsic/assets/id_utils.py
# Copyright 2023 Intrinsic Innovation LLC
# Intrinsic Proprietary and Confidential
# Provided subject to written agreement between the parties.

"""Provides utilities for working with asset ids."""
from __future__ import annotations

import dataclasses
import re

# These patterns differ from the C++ and Go versions in order to optimize
# pattern matching in Python. Simple matching on the patterns used in the other
# languages can be incredibly slow in Python (i.e., on the order of seconds for
# some single strings).

# The following match package, name, and id, respectively, except that:
# 1) They don't care about multiple underscores in a row.
# 2) They don't verify that package parts and names don't end with an
#    underscore.
# Matching should be done with _match_package, _match_name, and _match_id
# instead of these raw patterns.
_ALMOST_PACKAGE_PATTERN = re.compile(
    r"(?P<package>^([a-z][a-z0-9_]*\.)+[a-z][a-z0-9_]*$)"
)
_ALMOST_NAME_PATTERN = re.compile(r"(?P<name>^[a-z][a-z0-9_]*$)")
_ALMOST_ID_PATTERN = re.compile(
    r"(?P<id>^(?P<package>([a-z][a-z0-9_]*\.)+[a-z][a-z0-9_]*)\.(?P<name>[a-z][a-z0-9_]*)$)"
)

# Taken from semver.org.
_VERSION_PATTERN_STR = r"(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?"
_VERSION_PATTERN = re.compile(rf"^{_VERSION_PATTERN_STR}$")

# The id part of this pattern is not validated. It should be validated via
# `validate_id`.
_COARSE_ID_VERSION_PATTERN = re.compile(
    rf"^(?P<maybe_id>.+)\.(?P<version>{_VERSION_PATTERN_STR})$"
)

_LABEL_PATTERN = re.compile(r"^([a-z])|([a-z][a-z0-9-]*[a-z0-9])$")


class IdValidationError(ValueError):
  """An id_version or part of one is not valid."""


class CannotConvertLabelError(ValueError):
  """A string cannot be converted into a label."""


@dataclasses.dataclass(frozen=True)
class IdVersionParts:
  """Provides access to all of the parts of an id_version.

  See is_id_version for details about id_version formatting.

  Attributes:
    id: The id part of the id_version.
    id_version: The input id_version.
    name: The name part of the id_version.
    package: The package part of the id_version.
    version: The version of the id_version.
    version_build_metadata: The build metadata part of the version, if any.
    version_major: The major part of the version.
    version_minor: The minor part of the version.
    version_patch: The patch part of the version.
    version_pre_release: The pre-release part of the version, if any.
  """

  id: str
  id_version: str
  name: str
  package: str
  version: str
  version_build_metadata: str
  version_major: str
  version_minor: str
  version_patch: str
  version_pre_release: str

  @classmethod
  def create(cls, id_version: str) -> IdVersionParts:
    """Creates a new IdVersionParts from an id_version string.

    Args:
      id_version: The target id_version.

    Returns:
      The IdVersionParts for the specified id_version.

    Raises:
      IdValidationError: If the input id_version is not valid.
    """
    coarse_match, id_match = _match_id_version_or_raise(id_version)
    build_metadata = coarse_match.group("buildmetadata")
    pre_release = coarse_match.group("prerelease")

    return cls(
        id=id_match.group("id"),
        id_version=id_version,
        name=id_match.group("name"),
        package=id_match.group("package"),
        version=coarse_match.group("version"),
        version_build_metadata="" if build_metadata is None else build_metadata,
        version_major=coarse_match.group("major"),
        version_minor=coarse_match.group("minor"),
        version_patch=coarse_match.group("patch"),
        version_pre_release="" if pre_release is None else pre_release,
    )


def id_from(package: str, name: str) -> str:
  """Creates an id from package and name strings.

  Ids are formatted as in is_id.

  Args:
    package: The asset package.
    name: The asset name.

  Returns:
    The asset id.

  Raises:
    IdValidationError: If the specified package or name are not valid.
  """
  try:
    validate_package(package)
    validate_name(name)
  except IdValidationError as err:
    raise IdValidationError(
        f"Cannot create id_version from ({package}, {name})"
    ) from err

  return f"{package}.{name}"


def id_version_from(package: str, name: str, version: str) -> str:
  """Creates an id_version from package, name, and version strings.

  Id_versions are formatted as in is_id_version.

  Args:
    package: The asset package.
    name: The asset name.
    version: The asset version.

  Returns:
    The asset id_version.

  Raises:
    IdValidationError: If the specified package, name, or version are not valid.
  """
  try:
    validate_package(package)
    validate_name(name)
    validate_version(version)
  except IdValidationError as err:
    raise IdValidationError(
        f"Cannot create id_version from ({package}, {name}, {version})"
    ) from err

  return f"{package}.{name}.{version}"


def name_from(id: str) -> str:  # pylint: disable=redefined-builtin
  """Returns the name part of an id or id_version.

  Args:
    id: An id or id_version, as described in is_id and is_id_version,
      respectively.

  Returns:
    The package part of the id.

  Raises:
    IdValidationError: If the specified id is not valid.
  """
  match = _match_id(remove_version_from(id))
  if match is None:
    raise IdValidationError(f"{id!r} is not a valid id or id_version.")

  return match.group("name")


def package_from(id: str) -> str:  # pylint: disable=redefined-builtin
  """Returns the package part of an id or id_version.

  Args:
    id: An id or id_version, as described in is_id and is_id_version,
      respectively.

  Returns:
    The package part of the id.

  Raises:
    IdValidationError: If the specified id is not valid.
  """
  match = _match_id(remove_version_from(id))
  if match is None:
    raise IdValidationError(f"{id!r} is not a valid id or id_version.")

  return match.group("package")


def version_from(id_version: str) -> str:
  """Returns the version part of an id_version.

  Args:
    id_version: An id_version, as described in is_id_version.

  Returns:
    The version part of the id_version.

  Raises:
    IdValidationError: If the specified id_version is not valid.
  """
  coarse_match, _ = _match_id_version_or_raise(id_version)
  return coarse_match.group("version")


def remove_version_from(id: str) -> str:  # pylint: disable=redefined-builtin
  """Strips the version from `id` and returns the id substring.

  Args:
    id: An id or id_version, as described in is_id and is_id_version,
      respectively.

  Returns:
    The id part of the input with version, if any, stripped.

  Raises:
    IdValidationError: If the specified id is not valid.
  """
  try:
    _, id_match = _match_id_version_or_raise(id)
  except IdValidationError:
    validate_id(id)
    return id

  return id_match.group("id")


def is_id(id: str) -> bool:  # pylint: disable=redefined-builtin
  """Tests whether a string is a valid asset id.

  A valid id is formatted as "<package>.<name>", where `package` and `name` are
  formatted as described in is_package and is_name, respectively.

  Args:
    id: The string to test.

  Returns:
    True if `id` is a valid id.
  """
  return _match_id(id) is not None


def is_id_version(id_version: str) -> bool:
  """Tests whether a string is a valid asset id_version.

  A valid id_version is formatted as "<package>.<name>.<version>", where
  `package`, `name`, and `version` are formatted as described in is_package,
  is_name, and is_version, respectively.

  Args:
    id_version: The string to test.

  Returns:
    True if `id_version` is a valid id_version.
  """
  match = _COARSE_ID_VERSION_PATTERN.match(id_version)
  return match is not None and is_id(match.group("maybe_id"))


def is_name(name: str) -> bool:
  """Tests whether a string is a valid asset name.

  A valid name:
   - consists only of lower case alphanumeric characters and underscores;
   - begins with an alphabetic character;
   - ends with an alphanumeric character;
   - does not contain multiple underscores in a row.

  NOTE: Disallowing multiple underscores in a row enables underscores to be
  replaced with a hyphen (-) and periods to be replaced with two hyphens (--)
  in order to convert asset ids to labels (see `to_label`) without possibility
  of collisions.

  Args:
    name: The string to test.

  Returns:
    True if `name` is a valid name.
  """
  return _match_name(name) is not None


def is_package(package: str) -> bool:
  """Tests whether a string is a valid asset package.

  A valid package:
   - consists only of alphanumeric characters, underscores, and periods;
   - begins with an alphabetic character;
   - ends with an alphanumeric character;
   - contains at least one period;
   - precedes each period with an alphanumeric character;
   - follows each period with an alphabetic character;
   - does not contain multiple underscores in a row.

  NOTE: Disallowing multiple underscores in a row enables underscores to be
  replaced with a hyphen (-) and periods to be replaced with two hyphens (--)
  in order to convert asset ids to labels (see `to_label`) without possibility
  of collisions.

  Args:
    package: The string to test.

  Returns:
    True if `package` is a valid package.
  """
  return _match_package(package) is not None


def is_version(version: str) -> bool:
  """Tests whether a string is a valid asset version.

  A valid version is formatted as described by semver.org.

  Args:
    version: The string to test.

  Returns:
    True if `version` is a valid version.
  """
  return _VERSION_PATTERN.match(version) is not None


def validate_id(id: str) -> None:  # pylint: disable=redefined-builtin
  """Validates an id.

  A valid id is formatted as described in is_id.

  Args:
    id: The id to validate.

  Raises:
    IdValidationError: If the specified id is not valid.
  """
  if not is_id(id):
    raise IdValidationError(f"{id!r} is not a valid id.")


def validate_id_version(id_version: str) -> None:
  """Validates an id_version.

  A valid id_version is formatted as described in is_id_version.

  Args:
    id_version: The id_version to validate.

  Raises:
    IdValidationError: If the specified id_version is not valid.
  """
  if not is_id_version(id_version):
    raise IdValidationError(f"{id_version!r} is not a valid id_version.")


def validate_name(name: str) -> None:
  """Validates a name.

  A valid name is formatted as described in is_name.

  Args:
    name: The name to validate.

  Raises:
    IdValidationError: If the specified name is not valid.
  """
  if not is_name(name):
    raise IdValidationError(f"{name!r} is not a valid name.")


def validate_package(package: str) -> None:
  """Validates a package.

  A valid package is formatted as described in is_package.

  Args:
    package: The package to validate.

  Raises:
    IdValidationError: If the specified package is not valid.
  """
  if not is_package(package):
    raise IdValidationError(f"{package!r} is not a valid package.")


def validate_version(version: str) -> None:
  """Validates a version.

  A valid version is formatted as described in is_version.

  Args:
    version: The version to validate.

  Raises:
    IdValidationError: If the specified version is not valid.
  """
  if not is_version(version):
    raise IdValidationError(f"{version!r} is not a valid version.")


def parent_package_from(package: str) -> str:
  """Returns the parent package of the specified package.

  Returns an empty string if the package has no parent.

  NOTE: It does not validate the package.

  Args:
    package: The child package.

  Returns:
    The parent package, or an empty string if the child has no parent.
  """
  if package.count(".") < 2:
    return ""

  idx = package.rfind(".")
  return package[:idx]


def to_label(s: str) -> str:
  """Converts the input into a label.

  A label can be used as, e.g.:
    - a Kubernetes resource name
      (https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names);
    - a SpiceDB id (https://authzed.com/docs).

  A label:
    - consists of only alphanumeric characters and hyphens (-);
    - begins with an alphabetic character;
    - ends with an alphanumeric character.

  This function will potentially apply two transformations to the input:
    - "." is converted to "--";
    - "_" is converted to "-".

  If the above transformations cannot convert the input into a label, an error
  is raised.

  In order to support reversible transformations (see `from_label`), an input
  cannot be converted if it contains any of the following substrings: "-", "_.",
  "._", "__".

  Args:
    s: The string to convert.

  Returns:
    The converted label.

  Raises:
    CannotConvertLabelError: If the input cannot be converted into a label.
  """
  for offender in ("-", "_.", "._", "__"):
    if offender in s:
      raise CannotConvertLabelError(
          f"Cannot convert {s!r} into a label (contains {offender!r})."
      )

  label = s.replace("_", "-").replace(".", "--")

  if _LABEL_PATTERN.fullmatch(label) is None:
    raise CannotConvertLabelError(f"Cannot convert {s!r} into a label.")

  return label


def from_label(label: str) -> str:
  """Recovers an input string previously passed to `to_label`.

  Args:
    label: An output from `to_label`.

  Returns:
    The recovered string.
  """
  return label.replace("--", ".").replace("-", "_")


def _match_package(package: str) -> re.Match[str] | None:
  """Returns a regex Match for a package, or None for no match."""
  if "__" in package or "_." in package or package.endswith("_"):
    return None
  return _ALMOST_PACKAGE_PATTERN.match(package)


def _match_name(name: str) -> re.Match[str] | None:
  """Returns a regex Match for a name, or None for no match."""
  if "__" in name or name.endswith("_"):
    return None
  return _ALMOST_NAME_PATTERN.match(name)


def _match_id(id: str) -> re.Match[str] | None:  # pylint: disable=redefined-builtin
  """Returns a regex Match for an asset ID, or None for no match."""
  if "__" in id or "_." in id or id.endswith("_"):
    return None
  return _ALMOST_ID_PATTERN.match(id)


def _match_id_version_or_raise(
    id_version: str,
) -> tuple[re.Match[str], re.Match[str]]:
  """Returns regex Matches for an asset id_version, or raises for no match."""
  coarse_match = _COARSE_ID_VERSION_PATTERN.match(id_version)
  if coarse_match is None:
    raise IdValidationError(f"{id_version!r} is not a valid id_version.")

  id_match = _match_id(coarse_match.group("maybe_id"))
  if id_match is None:
    raise IdValidationError(f"{id_version!r} is not a valid id_version.")

  return coarse_match, id_match

File: intrinsic/assets/idutils.go
// Copyright 2023 Intrinsic Innovation LLC

// Package idutils contains utilities for interacting with id strings for assets (e.g., skills,
// resources).
//
// For example, it contains utilities to extract sub-components of the id strings, and check
// whether id strings, and their subcomponents (such as the version), are valid.
package idutils

import (
	"crypto/sha256"
	"errors"
	"fmt"
	"regexp"
	"strings"

	"golang.org/x/exp/slices"
	idpb "intrinsic/assets/proto/id_go_proto"
)

var (
	nameRegex    = regexp.MustCompile(`(?P<name>^[a-z]([a-z0-9_]?[a-z0-9])*$)`)
	packageRegex = regexp.MustCompile(`(?P<package>^([a-z]([a-z0-9_]?[a-z0-9])*\.)+([a-z]([a-z0-9_]?[a-z0-9])*)+$)`)
	// Taken from semver.org.
	versionRegex   = regexp.MustCompile(`(?P<version>^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$)`)
	idRegex        = regexp.MustCompile(`(?P<id>^(?P<package>([a-z]([a-z0-9_]?[a-z0-9])*\.)+[a-z]([a-z0-9_]?[a-z0-9])*)\.(?P<name>[a-z]([a-z0-9_]?[a-z0-9])*)$)`)
	idVersionRegex = regexp.MustCompile(`(?P<id_version>^(?P<id>(?P<package>([a-z]([a-z0-9_]?[a-z0-9])*\.)+[a-z]([a-z0-9_]?[a-z0-9])*)\.(?P<name>[a-z]([a-z0-9_]?[a-z0-9])*))\.(?P<version>(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)$)`)

	labelRegex          = regexp.MustCompile(`^[a-z]([a-z0-9\-]*[a-z0-9])*$`)
	labelFirstCharRegex = regexp.MustCompile(`^[a-z].*`)
	labelBadCharRegex   = regexp.MustCompile(`[^a-z0-9\-]`)
	labelLastCharRegex  = regexp.MustCompile(`.*[a-z0-9]$`)

	nonReleasedVersionRegex = regexp.MustCompile("\\+(?:sideloaded|inlined)")
)

// getNamedMatch extracts the named group from a match of a string on a regex pattern.
func getNamedMatch(str string, re *regexp.Regexp, group string) (string, error) {
	groups := re.SubexpNames()
	submatches := re.FindStringSubmatch(str)
	if submatches == nil {
		return "", fmt.Errorf("%q is not a valid %s", str, groups[1])
	}

	groupIdx := slices.Index(groups, group)
	if groupIdx == -1 {
		return "", fmt.Errorf("unknown group: %q (groups: %v)", group, groups)
	}

	return submatches[groupIdx], nil
}

// IDVersionParts provides access to all of the parts of an id_version.
//
// See IsIDVersion for details about id_version formatting.
type IDVersionParts struct {
	id                   string
	idVersion            string
	name                 string
	pkg                  string
	version              string
	versionBuildMetadata string
	versionMajor         string
	versionMinor         string
	versionPatch         string
	versionPreRelease    string
}

// NewIDVersionParts creates a new IDVersionParts from an id_version string.
func NewIDVersionParts(idVersion string) (*IDVersionParts, error) {
	submatches := idVersionRegex.FindStringSubmatch(idVersion)
	if submatches == nil {
		return nil, fmt.Errorf("%q is not a valid id_version", idVersion)
	}

	groups := idVersionRegex.SubexpNames()

	return &IDVersionParts{
		id:                   submatches[slices.Index(groups, "id")],
		idVersion:            idVersion,
		name:                 submatches[slices.Index(groups, "name")],
		pkg:                  submatches[slices.Index(groups, "package")],
		version:              submatches[slices.Index(groups, "version")],
		versionBuildMetadata: submatches[slices.Index(groups, "buildmetadata")],
		versionMajor:         submatches[slices.Index(groups, "major")],
		versionMinor:         submatches[slices.Index(groups, "minor")],
		versionPatch:         submatches[slices.Index(groups, "patch")],
		versionPreRelease:    submatches[slices.Index(groups, "prerelease")],
	}, nil
}

// ID returns the id part of id_version.
func (p *IDVersionParts) ID() string {
	return p.id
}

// IDProto returns the id proto part of id_version.
func (p *IDVersionParts) IDProto() *idpb.Id {
	return &idpb.Id{Package: p.pkg, Name: p.name}
}

// IDVersion returns the id_version.
func (p *IDVersionParts) IDVersion() string {
	return p.idVersion
}

// IDVersionProto returns the id_version proto.
func (p *IDVersionParts) IDVersionProto() *idpb.IdVersion {
	return &idpb.IdVersion{Id: p.IDProto(), Version: p.Version()}
}

// Name returns the name part of id_version.
func (p *IDVersionParts) Name() string {
	return p.name
}

// Package returns the package part of id_version.
func (p *IDVersionParts) Package() string {
	return p.pkg
}

// Version returns the version part of id_version.
func (p *IDVersionParts) Version() string {
	return p.version
}

// VersionBuildMetadata returns the build metadata part of id_version.
func (p *IDVersionParts) VersionBuildMetadata() string {
	return p.versionBuildMetadata
}

// VersionMajor returns the major part of the version.
func (p *IDVersionParts) VersionMajor() string {
	return p.versionMajor
}

// VersionMinor returns the minor part of the version.
func (p *IDVersionParts) VersionMinor() string {
	return p.versionMinor
}

// VersionPatch returns the patch part of the version.
func (p *IDVersionParts) VersionPatch() string {
	return p.versionPatch
}

// VersionPreRelease returns the pre-release part of the version, if any.
func (p *IDVersionParts) VersionPreRelease() string {
	return p.versionPreRelease
}

// IDFrom creates an id from package and name strings.
//
// Ids are formatted as in IsId.
//
// Returns an error if `pkg` or `name` strings not valid.
func IDFrom(pkg string, name string) (string, error) {
	err := ValidatePackage(pkg)
	if err == nil {
		err = ValidateName(name)
	}
	if err != nil {
		return "", fmt.Errorf("cannot create id from (%q, %q): %v", pkg, name, err)
	}

	return fmt.Sprintf("%s.%s", pkg, name), nil
}

// IDProtoFrom creates an Id proto from package and name strings.
//
// Returns an error if `pkg` or `name` strings not valid.
func IDProtoFrom(pkg string, name string) (*idpb.Id, error) {
	err := ValidatePackage(pkg)
	if err == nil {
		err = ValidateName(name)
	}
	if err != nil {
		return nil, fmt.Errorf("cannot create Id from (%q, %q): %v", pkg, name, err)
	}

	return &idpb.Id{Name: name, Package: pkg}, nil
}

// IDFromProto creates an id string from an Id proto message.
//
// Ids are formatted as in IsId.
//
// Returns an error if `package` or `name` fields are not valid.
func IDFromProto(id *idpb.Id) (string, error) {
	return IDFrom(id.GetPackage(), id.GetName())
}

// IDVersionFrom creates an id_version from package, name, and version strings.
//
// Id_versions are formatted as in IsIdVersion.
//
// Returns an error if `pkg`, `name`, or `version` strings are not valid.
func IDVersionFrom(pkg string, name string, version string) (string, error) {
	err := ValidatePackage(pkg)
	if err == nil {
		err = ValidateName(name)
	}
	if err == nil {
		err = ValidateVersion(version)
	}
	if err != nil {
		return "", fmt.Errorf("cannot create id_version from (%q, %q, %q): %v", pkg, name, version, err)
	}

	return fmt.Sprintf("%s.%s.%s", pkg, name, version), nil
}

// IDVersionProtoFrom creates an IdVersion proto from package, name, and version strings.
//
// Returns an error if `pkg`, `name`, or `version` strings are not valid.
func IDVersionProtoFrom(pkg string, name string, version string) (*idpb.IdVersion, error) {
	id, err := IDProtoFrom(pkg, name)
	if err != nil {
		return nil, err
	}
	if err := ValidateVersion(version); err != nil {
		return nil, fmt.Errorf("cannot create IdVersion from (%q, %q, %q): %v", pkg, name, version, err)
	}

	return &idpb.IdVersion{Id: id, Version: version}, nil
}

// IDVersionFromProto creates an id_version string from an IdVersion proto message.
//
// Id_versions are formatted as in IsIdVersion.
//
// Returns an error if `package`, `name`, or `version` fields are not valid.
func IDVersionFromProto(idVersion *idpb.IdVersion) (string, error) {
	return IDVersionFrom(idVersion.GetId().GetPackage(), idVersion.GetId().GetName(), idVersion.GetVersion())
}

// NameFrom returns the name part of an id or id_version.
//
// `id` must be formatted as an id or id_version, as described in IsId and IsIdVersion,
// respectively.
func NameFrom(id string) (string, error) {
	name, err := getNamedMatch(id, idVersionRegex, "name")
	if err == nil {
		return name, nil
	}

	return getNamedMatch(id, idRegex, "name")
}

// PackageFrom returns the package part of an id or id_version.
//
// `id` must be formatted as an id or id_version, as described in IsId and IsIdVersion,
// respectively.
func PackageFrom(id string) (string, error) {
	pkg, err := getNamedMatch(id, idVersionRegex, "package")
	if err == nil {
		return pkg, nil
	}

	return getNamedMatch(id, idRegex, "package")
}

// VersionFrom returns the  version part of an id_version.
//
// `id_version` must be formatted as described in IsIdVersion.
func VersionFrom(idVersion string) (string, error) {
	return getNamedMatch(idVersion, idVersionRegex, "version")
}

// RemoveVersionFrom strips the version from `id` and returns the id substring.
//
// `id` must be formatted as an id or id_version, as described in IsId and IsIdVersion,
// respectively.
//
// If there is no version information in the given `id`, the returned value will equal `id`.
func RemoveVersionFrom(id string) (string, error) {
	stripped, err := getNamedMatch(id, idVersionRegex, "id")
	if err == nil {
		return stripped, nil
	}

	if err := ValidateID(id); err != nil {
		return "", err
	}
	return id, nil
}

// IsID Tests whether a string is a valid asset id.
//
// A valid id is formatted as "<package>.<name>", where `package` and `name` are formatted as
// described in IsPackage and IsName, respectively.
func IsID(id string) bool {
	return idRegex.MatchString(id)
}

// IsIDVersion tests whether a string is a valid asset id_version.
//
// A valid id_version is formatted as "<package>.<name>.<version>", where `package`, `name`, and
// `version` are formatted as described in IsPackage, IsName, and IsVersion, respectively.
func IsIDVersion(idVersion string) bool {
	return idVersionRegex.MatchString(idVersion)
}

// IsName tests whether a string is a valid asset name.
//
// A valid name:
//   - consists only of lower case alphanumeric characters and underscores;
//   - begins with an alphabetic character;
//   - ends with an alphanumeric character;
//   - does not contain multiple underscores in a row.
//
// NOTE: Disallowing multiple underscores in a row enables underscores to be replaced with a hyphen
// (-) and periods to be replaced with two hyphens (--) in order to convert asset ids to kubernetes
// labels without possibility of collisions.
func IsName(name string) bool {
	return nameRegex.MatchString(name)
}

// IsPackage tests whether a string is a valid asset package.
//
// A valid package:
//   - consists only of alphanumeric characters, underscores, and periods;
//   - begins with an alphabetic character;
//   - ends with an alphanumeric character;
//   - contains at least one period;
//   - precedes each period with an alphanumeric character;
//   - follows each period with an alphabetic character;
//   - does not contain multiple underscores in a row.
//
// NOTE: Disallowing multiple underscores in a row enables underscores to be replaced with a hyphen
// (-) and periods to be replaced with two hyphens (--) in order to convert asset ids to kubernetes
// labels without possibility of collisions.
func IsPackage(pkg string) bool {
	return packageRegex.MatchString(pkg)
}

// IsVersion tests whether a string is a valid asset version.
//
// A valid version is formatted as described by semver.org.
func IsVersion(version string) bool {
	return versionRegex.MatchString(version)
}

// IsUnreleasedVersion tests whether a string is a valid and unreleased asset version.
//
// A valid unreleased version is formatted as described by semver.org with build metadata matching
// the reserved prefix for unreleased assets.
func IsUnreleasedVersion(version string) bool {
	return IsVersion(version) && nonReleasedVersionRegex.MatchString(version)
}

// ValidateID validates an id.
//
// A valid id is formatted as described in IsId.
//
// Returns an error if `id` is not valid.
func ValidateID(id string) error {
	if !IsID(id) {
		return fmt.Errorf("%q is not a valid id", id)
	}
	return nil
}

// ValidateIDProto validates the parts of an Id proto.
//
// Returns an error if `idProto` is not valid.
func ValidateIDProto(idProto *idpb.Id) error {
	if err := ValidatePackage(idProto.GetPackage()); err != nil {
		return err
	}
	if err := ValidateName(idProto.GetName()); err != nil {
		return err
	}

	return nil
}

// ValidateIDVersion validates an id_version.
//
// A valid id_version is formatted as described in IsIdVersion.
//
// Returns an error if `idVersion` is not valid.
func ValidateIDVersion(idVersion string) error {
	if !IsIDVersion(idVersion) {
		return fmt.Errorf("%q is not a valid id_version", idVersion)
	}
	return nil
}

// ValidateName validates a name.
//
// A valid name is formatted as described in IsName.
//
// Returns an error if `name` is not valid.
func ValidateName(name string) error {
	if !IsName(name) {
		return fmt.Errorf("%q is not a valid name", name)
	}
	return nil
}

// ValidatePackage validates a package.
//
// A valid package is formatted as described in IsPackage.
//
// Returns an error if `pkg` is not valid.
func ValidatePackage(pkg string) error {
	if !IsPackage(pkg) {
		return fmt.Errorf("%q is not a valid package", pkg)
	}
	return nil
}

// ValidateVersion validates a version.
//
// A version is formatted as described in IsVersion.
//
// Returns an error if `version` is not valid.
func ValidateVersion(version string) error {
	if !IsVersion(version) {
		return fmt.Errorf("%q is not a valid version", version)
	}
	return nil
}

// ValidateUnreleasedVersion validates an unreleased version.
//
// An unreleased version is formatted as described in IsUnreleasedVersion.
//
// Returns an error if `version` is not valid or not unreleased.
func ValidateUnreleasedVersion(version string) error {
	if !IsUnreleasedVersion(version) {
		return fmt.Errorf("%q is not a valid unreleased version", version)
	}
	return nil
}

// ParentFromPackage returns the parent package of the specified package/
//
// Returns an empty string if the package has no parent.
//
// NOTE: It does not validate the package.
func ParentFromPackage(pkg string) string {
	if n := strings.Count(pkg, "."); n < 2 {
		return "" // No parent.
	}

	idx := strings.LastIndex(pkg, ".")
	return pkg[:idx]
}

// ToLabel converts the input into a label.
//
// A label can be used as, e.g.:
//   - a Kubernetes resource name
//     (https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names);
//   - a SpiceDB id (https://authzed.com/docs).
//
// A label:
//   - consists of only lower case alphanumeric characters and hyphens (-);
//   - begins with an alphabetic character;
//   - ends with an alphanumeric character.
//
// This function will potentially apply two transformations to the input:
//   - "." is converted to "--";
//   - "_" is converted to "-".
//
// If the above transformations cannot convert the input into a label, an error is returned.
//
// In order to support reversible transformations (see `FromLabel`), an input cannot be converted if
// it contains any of the following substrings: "-", "_.", "._", "__".
func ToLabel(s string) (string, error) {
	for _, offender := range []string{"-", "_.", "._", "__"} {
		if strings.Contains(s, offender) {
			return "", fmt.Errorf("cannot convert %q into a label (contains %q)", s, offender)
		}
	}

	label := strings.ReplaceAll(strings.ReplaceAll(s, "_", "-"), ".", "--")

	if !labelRegex.MatchString(label) {
		return "", fmt.Errorf("cannot convert %q into a label (got invalid label: %q)", s, label)
	}

	return label, nil
}

// ToLabelNonReversible converts the input into a label (see ToLabel).
//
// The label may not be reversible using FromLabel.
func ToLabelNonReversible(s string) (string, error) {
	if len(s) == 0 {
		return "", fmt.Errorf("cannot convert empty string into a label")
	}
	label := strings.ToLower(s)

	label = strings.ReplaceAll(strings.ReplaceAll(label, "_", "-"), ".", "--")
	label = labelBadCharRegex.ReplaceAllString(label, "-")
	if !labelFirstCharRegex.MatchString(label) {
		label = fmt.Sprintf("a%s", label)
	}
	if !labelLastCharRegex.MatchString(label) {
		label = fmt.Sprintf("%sa", label)
	}

	if !labelRegex.MatchString(label) {
		return "", fmt.Errorf("cannot convert %q into a label (got invalid label: %q)", s, label)
	}

	return label, nil
}

// FromLabel recovers an input string previously passed to ToLabel.
func FromLabel(label string) string {
	return strings.ReplaceAll(strings.ReplaceAll(label, "--", "."), "-", "_")
}

// UnreleasedAssetKind describes the kind (i.e. the "source") of an unreleased asset.
//
// The kind is encoded in the version of the asset.
type UnreleasedAssetKind string

const (
	// UnreleasedAssetKindSideloaded describes an asset that has been or will be sideloaded.
	UnreleasedAssetKindSideloaded = "sideloaded"

	// UnreleasedAssetKindInlined describes an asset that has been or will be inlined from source.
	UnreleasedAssetKindInlined = "inlined"
)

// UnreleasedVersion creates a deterministic, unreleased version of the given kind. The version is
// marked with a digest of the data to provide strong uniqueness guarantees.
//
// The resulting version is formatted as described in IsUnreleasedVersion.
//
// Returns an error if the kind is unknown/invalid or data is empty.
func UnreleasedVersion(kind UnreleasedAssetKind, data []byte) (string, error) {
	if kind != UnreleasedAssetKindSideloaded && kind != UnreleasedAssetKindInlined {
		return "", errors.New("invalid kind")
	}
	if len(data) == 0 {
		return "", errors.New("empty data")
	}
	return fmt.Sprintf("0.0.1+%s%x", kind, sha256.Sum256(data)), nil
}
